<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en_US" lang="en_US">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<!-- qexception.cpp -->
  <title>QException Class | QtCore 5.3</title>
  <link rel="stylesheet" type="text/css" href="style/offline.css" />
</head>
<body>
<div class="header" id="qtdocheader">
    <div class="main">
    <div class="main-rounded">
        <div class="navigationbar">
        <ul>
<li><a href="../qtdoc/index.html">Qt 5.3</a></li>
<li><a href="qtcore-index.html">Qt Core</a></li>
<li><a href="qtcore-module.html">C++ Classes</a></li>
<li>QException</li>
<li id="buildversion">
Qt 5.3.1 Reference Documentation</li>
    </ul>
    </div>
</div>
<div class="content">
<div class="line">
<div class="content mainContent">
<div class="toc">
<h3><a name="toc">Contents</a></h3>
<ul>
<li class="level1"><a href="#public-functions">Public Functions</a></li>
<li class="level1"><a href="#details">Detailed Description</a></li>
</ul>
</div>
<h1 class="title">QException Class</h1>
<!-- $$$QException-brief -->
<p>The QException class provides a base class for exceptions that can transferred across threads. <a href="#details">More...</a></p>
<!-- @@@QException -->
<table class="alignedsummary">
<tr><td class="memItemLeft rightAlign topAlign"> Header:</td><td class="memItemRight bottomAlign"> </b><tt><span class="preprocessor">#include &lt;QException&gt;</span>
</tt></td></tr><tr><td class="memItemLeft rightAlign topAlign"> qmake:</td><td class="memItemRight bottomAlign"> <tt>QT += core</tt></td></tr><tr><td class="memItemLeft rightAlign topAlign"> Since:</td><td class="memItemRight bottomAlign">  Qt 5.0</td></tr><tr><td class="memItemLeft rightAlign topAlign"> Inherited By:</td><td class="memItemRight bottomAlign"> <p><a href="qunhandledexception.html">QUnhandledException</a>.</p>
</td></tr></table><ul>
<li><a href="qexception-members.html">List of all members, including inherited members</a></li>
</ul>
<a name="public-functions"></a>
<h2>Public Functions</h2>
<table class="alignedsummary">
<tr><td class="memItemLeft rightAlign topAlign"> virtual QException * </td><td class="memItemRight bottomAlign"><b><a href="qexception.html#clone">clone</a></b>() const</td></tr>
<tr><td class="memItemLeft rightAlign topAlign"> virtual void </td><td class="memItemRight bottomAlign"><b><a href="qexception.html#raise">raise</a></b>() const</td></tr>
</table>
<a name="details"></a>
<!-- $$$QException-description -->
<div class="descr">
<h2>Detailed Description</h2>
<p>The QException class provides a base class for exceptions that can transferred across threads.</p>
<p>Qt Concurrent supports throwing and catching exceptions across thread boundaries, provided that the exception inherit from QException and implement two helper functions:</p>
<pre class="cpp"><span class="keyword">class</span> MyException : <span class="keyword">public</span> <span class="type">QException</span>
{
<span class="keyword">public</span>:
    <span class="type">void</span> raise() <span class="keyword">const</span> { <span class="keyword">throw</span> <span class="operator">*</span><span class="keyword">this</span>; }
    MyException <span class="operator">*</span>clone() <span class="keyword">const</span> { <span class="keyword">return</span> <span class="keyword">new</span> MyException(<span class="operator">*</span><span class="keyword">this</span>); }
};</pre>
<p>QException subclasses must be thrown by value and caught by reference:</p>
<pre class="cpp"><span class="keyword">try</span>  {
    <span class="type"><a href="../qtconcurrent/qtconcurrent.html">QtConcurrent</a></span><span class="operator">::</span>blockingMap(list<span class="operator">,</span> throwFunction); <span class="comment">// throwFunction throws MyException</span>
} <span class="keyword">catch</span> (MyException <span class="operator">&amp;</span>e) {
    <span class="comment">// handle exception</span>
}</pre>
<p>If you throw an exception that is not a subclass of QException, the Qt functions will throw a <a href="qunhandledexception.html">QUnhandledException</a> in the receiver thread.</p>
<p>When using <a href="qfuture.html">QFuture</a>, transferred exceptions will be thrown when calling the following functions:</p>
<ul>
<li><a href="qfuture.html#waitForFinished">QFuture::waitForFinished</a>()</li>
<li><a href="qfuture.html#result">QFuture::result</a>()</li>
<li><a href="qfuture.html#resultAt">QFuture::resultAt</a>()</li>
<li><a href="qfuture.html#results">QFuture::results</a>()</li>
</ul>
</div>
<!-- @@@QException -->
<div class="func">
<h2>Member Function Documentation</h2>
<!-- $$$clone[overload1]$$$clone -->
<h3 class="fn"><a name="clone"></a><span class="type">QException</span> * QException::<span class="name">clone</span>() const<tt> [virtual]</tt></h3>
<p>In your <a href="qexception.html">QException</a> subclass, reimplement clone() like this:</p>
<pre class="cpp">MyException <span class="operator">*</span>MyException<span class="operator">::</span>clone() <span class="keyword">const</span> { <span class="keyword">return</span> <span class="keyword">new</span> MyException(<span class="operator">*</span><span class="keyword">this</span>); }</pre>
<!-- @@@clone -->
<!-- $$$raise[overload1]$$$raise -->
<h3 class="fn"><a name="raise"></a><span class="type">void</span> QException::<span class="name">raise</span>() const<tt> [virtual]</tt></h3>
<p>In your <a href="qexception.html">QException</a> subclass, reimplement raise() like this:</p>
<pre class="cpp"><span class="type">void</span> MyException<span class="operator">::</span>raise() <span class="keyword">const</span> { <span class="keyword">throw</span> <span class="operator">*</span><span class="keyword">this</span>; }</pre>
<!-- @@@raise -->
</div>
        </div>
       </div>
   </div>
   </div>
</div>
<div class="footer">
   <p>
   <acronym title="Copyright">&copy;</acronym> 2014 Digia Plc and/or its
   subsidiaries. Documentation contributions included herein are the copyrights of
   their respective owners.<br>    The documentation provided herein is licensed under the terms of the    <a href="http://www.gnu.org/licenses/fdl.html">GNU Free Documentation    License version 1.3</a> as published by the Free Software Foundation.<br>    Digia, Qt and their respective logos are trademarks of Digia Plc     in Finland and/or other countries worldwide. All other trademarks are property
   of their respective owners. </p>
</div>
</body>
</html>
